### PR TITLE
feat(Haptics): create haptic pattern based on AudioSource

### PR DESCRIPTION
**Pull request by fight4dream**
This haptic generated reflects the current playing audio even if
the audio is looping, delayed, volume/pitch modulated.

### CODE REVIEW

File: Runtime/Haptics/AudioSourceHapticPulser.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            }
+
+            StopCoroutine(hapticRoutine);
+            hapticRoutine = null;
+            HapticPulser.Cancel();
+            ResetIntensity();
+        }
+
+        /// <summary>
+        /// Resets the <see cref="Haptics.HapticPulser.Intensity"/> back to it's original value.
```
Comment (bddckr) [inline Runtime/Haptics/AudioSourceHapticPulser.cs]:
```suggestion
        /// Resets the <see cref="Haptics.HapticPulser.Intensity"/> back to its original value.
```

**Comment by bddckr** [comment]
Also: I _looove_ this idea. 💓

**Comment by fight4dream** [comment]
base class name `AudioHapticPulser` ?

**Comment by fight4dream** [comment]
Oh i guess i overlooked it. 
Btw, any suggestions on the commit title/message?

**Comment by bddckr** [comment]
Looking good to me, but perhaps should mention the refactoring changes:
```
feat(Haptics): create haptic pattern based on AudioSource

This haptic generated reflects the current playing audio even if the
audio is looping, delayed, volume/pitch modulated. Since this new
implementation shares the same core logic as the existing
AudioClipHapticPulser the shared logic has been extracted into an
abstract superclass.
```

File: Runtime/Haptics/AudioClipHapticPulser.cs [hunk 2]
```diff
@@ -1,91 +1,42 @@
 ﻿namespace Zinnia.Haptics
 {
-    using UnityEngine;
     using System.Collections;
-    using Malimbe.XmlDocumentationAttribute;
+    using UnityEngine;
```
Comment (thestonefox) [inline Runtime/Haptics/AudioClipHapticPulser.cs]: As before, we put `Unity*` before `System*` (for now)

File: Runtime/Haptics/AudioSourceHapticPulser.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+﻿namespace Zinnia.Haptics
+{
+    using System.Collections;
```
Comment (thestonefox) [inline Runtime/Haptics/AudioSourceHapticPulser.cs]: apply consistent namespace order again

File: Runtime/Haptics/RoutineHapticPulser.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
+﻿namespace Zinnia.Haptics
+{
+    using System.Collections;
```
Comment (thestonefox) [inline Runtime/Haptics/RoutineHapticPulser.cs]: and here

**Comment by thestonefox** [comment]
I've tested this in Unity and I only get a single buzz for the first frame, it's not following the audioclip pattern.

I would assume this set up:
image

Would give me the same output as this set up:
image

**Comment by fight4dream** [comment]
The usage for audio source haptics is 
1. AudioSource.Play() or other methods to make it playing
2. Then call haptics.Begin()

The first buzz is actually a bug, if the audio source isnt playing, it shouldnt buzz at all and should immediately end the routine

**Comment by thestonefox** [comment]
Actually i figured out what i was doing wrong. I got it to work, but then after the loop I got 

```
IndexOutOfRangeException: Index was outside the bounds of the array.
Zinnia.Haptics.AudioSourceHapticPulser+<HapticProcessRoutine>d__7.MoveNext () (at Assets/Zinnia.Unity/Runtime/Haptics/AudioSourceHapticPulser.cs:50)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at C:/buildslave/unity/build/Runtime/Export/Scripting/Coroutines.cs:17)
```

**Comment by fight4dream** [comment]
Ok i will have it fix soon. Any comment on its usage for now?
I wish there is a Played event from audio source but nope

**Comment by fight4dream** [comment]
i've changed it to use OnAudioFilterRead()
hence, this audio source haptic pulser need to be placed after audio source as well as other audio filters

**Comment by thestonefox** [comment]
Not sure if @bddckr wants to have a final review. If not, then I'm happy to merge it

**Pull request merged**